### PR TITLE
Move Client.Watch inside Client.Attach

### PR DESCRIPTION
**Pull request by krapie**
**What this PR does / why we need it**:

Move `Client.Watch()` inside `Client.Attach()` to hide it from external interface.
This PR also introduces `Client.Subscribe()` which performs similar functionality with other SDKs. 

This PR is recreated from https://github.com/yorkie-team/yorkie/pull/593 

**Which issue(s) this PR fixes**:

Address #584

**Special notes for your reviewer**:

I prefer keeping this PR size as small as possible to wrap up this task on time.

**Does this PR introduce a user-facing change?**:

```release-note
Go SDK user will use `Client.Subscribe()` instead of `Client.Watch()`, just like other SDKs.
```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

### CODE REVIEW

**Comment by hackerwins** [comment]
Thank you for your contribution. In JS SDK, there is also a setting for SyncMode as shown in the link below. How about applying them together?

https://github.com/yorkie-team/yorkie-js-sdk/pull/772#pullrequestreview-1988728945

**Comment by krapie** [comment]
Is the syncMode feature big? Like I mentioned, I want to keep this PR small.

File: client/client.go [hunk 1]
```diff
@@ -406,21 +428,37 @@ func (c *Client) Sync(ctx context.Context, options ...SyncOptions) error {
...
+) (<-chan WatchResponse, context.CancelFunc, error) {
+	attachment, ok := c.attachments[doc.Key()]
+	if !ok {
+		return nil, nil, ErrDocumentNotAttached
+	}
+
+	if attachment.isSubscribed.Load() {
+		return nil, nil, ErrAlreadySubscribed
+	}
+	attachment.isSubscribed.Store(true)
```
Comment (hackerwins) [inline client/client.go:443]: I think this employs the "check-then-act" pattern, where it first checks the state and then changes it. I suspect that there could be a race condition.

How about introducing CompareAndSet, which can perform state comparison and setting atomically?

https://pkg.go.dev/sync/atomic#Bool.CompareAndSwap

File: client/client.go [hunk 2]
```diff
@@ -72,12 +73,22 @@ var (
...
+	ErrAlreadySubscribed = errors.New("already subscribed")
 )
 
 // Attachment represents the document attached.
 type Attachment struct {
 	doc   *document.Document
 	docID types.ID
+
+	// TODO(krapie): assuming that a client do not open multiple subscriptions for the same document.
+	isSubscribed atomic.Bool
```
Comment (hackerwins) [inline client/client.go:87]: This comment is `TODO`, so how about writing about what we think will happen in the future?

**Review by hackerwins — CHANGES_REQUESTED** [review]
Thanks for your contribution. 🙏
I left simple questions.

File: client/client.go (hunk 1 above)
Comment (krapie) [reply client/client.go:443]: Oops, this will definitely create race condition. Thank you for your feedback :)

**Review by hackerwins — APPROVED** [review]
LGTM. 👍

**Comment by hackerwins** [comment]
@krapie, @karockai Thank you for your contribution.

**Pull request merged**